### PR TITLE
adds basic checking of subscription state

### PR DESCRIPTION
**Pull request by s-amann**
### What this PR does
**Before this PR:**
Subscription state was not validated when making requests belonging to a subscription

**After this PR:**
Subscription state is validated, meaning, requests must be using a subscriptionID which has a 'registered' state stored in the frontends cache.

To add any subscription as registered to the frontend cache run

```bash
curl -X PUT localhost:8443/subscriptions/SUBSCRIPTION_ID?api-version=2.0 -H 'content-type: application/json' -d '{"state":"Registered"}'
```

Jira: ARO-6313
Link to demo recording: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Deployment: The deployment process was considered and addressed if required
- [x] Testing: New code requires new unit tests.
- [x] Documentation: Is the documentation updated? Either in the doc located in focus area, in the README or in the code itself.
- [x] Customers: Is this change affecting customers? Is the release plan considered?

### CODE REVIEW

File: frontend/middleware_validatesubscription.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	cache *Cache
+}
+
+func NewSubscriptionStateMuxValidator(c *Cache) *SubscriptionStateMuxValidator {
+	return &SubscriptionStateMuxValidator{
+		cache: c,
+	}
+}
+
+func (s *SubscriptionStateMuxValidator) MiddlewareValidateSubscriptionState(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
```
Comment (AldoFusterTurpin) [inline frontend/middleware_validatesubscription.go:26]: Maybe we can name this method just _ValidateSubscriptionState_ to avoid redundancy.

Comment (AldoFusterTurpin) [inline frontend/middleware_validatesubscription.go:26]: We could add a "return" in each of the if/else branches of the functions. We would convert the else's to simple ifs and avoid nesting the levels. WDT? 🙂

**Review by AldoFusterTurpin — COMMENTED** [review]
I know it is in draft mode, but just two minor suggestions in case they make sense to you. Thanks! 🙂

File: frontend/middleware_validatesubscription.go (hunk 1 above)
Comment (s-amann) [reply frontend/middleware_validatesubscription.go:26]: yeah, this logic is a mess, just a rough outline of what needs to get checked. Returning early is something I'll add in :)

Comment (s-amann) [reply frontend/middleware_validatesubscription.go:35]: I prefer to keep `Middleware` in the name, so it aligns with all other `Middleware`

File: frontend/middleware_validatesubscription.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	"strings"
+
+	"github.com/Azure/ARO-HCP/internal/api"
+	"github.com/Azure/ARO-HCP/internal/api/arm"
+	"github.com/Azure/ARO-HCP/internal/api/subscription"
+)
+
+type SubscriptionStateMuxValidator struct {
+	cache *Cache
+}
```
Comment (mbarnes) [inline frontend/middleware_validatesubscription.go:25]: Suggestion...

If we added the cache to the `http.Server`'s base context in `NewFrontend` under a new `ContextKeyCache` constant, then this middleware wouldn't need its own struct (which is cleaner, in my opinion).  `MiddlewareValidateSubscriptionState` could just fetch the cache from the given request's context.

File: frontend/middleware_validatesubscription.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+func NewSubscriptionStateMuxValidator(c *Cache) *SubscriptionStateMuxValidator {
+	return &SubscriptionStateMuxValidator{
+		cache: c,
+	}
+}
+
+func (s *SubscriptionStateMuxValidator) MiddlewareValidateSubscriptionState(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	subscriptionId := r.PathValue(strings.ToLower(PathSegmentSubscriptionID))
```
Comment (mbarnes) [inline frontend/middleware_validatesubscription.go:36]: Is lowercasing the path segment name going break the lookup?

File: frontend/middleware_validatesubscription.go [hunk 4]
```diff
@@ -23,44 +30,54 @@ func NewSubscriptionStateMuxValidator(c *Cache) *SubscriptionStateMuxValidator {
 	}
 }
 
+// MiddlewhareValidateSubscriptionState validates the state of the subscription as outlined by
+// https://github.com/cloud-and-ai-microsoft/resource-provider-contract/blob/master/v1.0/subscription-lifecycle-api-reference.md
```
Comment (mbarnes) [inline frontend/middleware_validatesubscription.go:34]: :+1: for the documentation link.

**Review by mbarnes — DISMISSED** [review]
Had a suggestion that would require a bit of refactoring, so I'll leave it up to you.  LGTM otherwise.

File: frontend/middleware_validatesubscription.go (hunk 1 above)
Comment (AldoFusterTurpin) [reply frontend/middleware_validatesubscription.go:35]: Makes sense

File: frontend/middleware_validatesubscription.go (hunk 3 above)
Comment (s-amann) [reply frontend/middleware_validatesubscription.go:36]: This middleware will be done after the `lowercase` middleware, since it is in the `postMux` flow.

Comment (mbarnes) [reply frontend/middleware_validatesubscription.go:36]: Should it be `strings.ToLower(r.PathValue(PathSegmentSubscriptionID))` then?  The path value key isn't lowercased, only the value.

File: frontend/middleware_validatesubscription.go (hunk 2 above)
Comment (mbarnes) [reply frontend/middleware_validatesubscription.go:25]: That's a fair point, I'm not sure we would want that either.  Ok, so good as is.

File: frontend/frontend.go [hunk 5]
```diff
@@ -69,11 +71,14 @@ func NewFrontend(logger *slog.Logger, listener net.Listener) *Frontend {
 	// Unauthenticated routes
 	mux.HandleFunc("/", f.NotFound)
 	mux.HandleFunc(MuxPattern(http.MethodGet, "healthz", "ready"), f.HealthzReady)
+	// TODO: determine where in the auth chain we should allow for this endpoint to be called by ARM
+	mux.HandleFunc(MuxPattern(http.MethodPut, PatternSubscriptions), f.ArmSubscriptionAction)
```
Comment (bennerv) [inline frontend/frontend.go:75]: TODO: do we have a card for MISE integration for P3?

File: frontend/middleware_validatesubscription.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	"net/http"
+	"strings"
+
+	"github.com/Azure/ARO-HCP/internal/api"
+	"github.com/Azure/ARO-HCP/internal/api/arm"
+	"github.com/Azure/ARO-HCP/internal/api/subscription"
+)
+
+const (
+	UnregisteredSubscriptionStateMessage = "The subscription %s is not registered for this provider %s. Please re-register the subscription."
```
Comment (bennerv) [inline frontend/middleware_validatesubscription.go:17]: are these messages the same as ARO Classic?

File: frontend/middleware_validatesubscription.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+		cache: c,
+	}
+}
+
+// MiddlewareValidateSubscriptionState validates the state of the subscription as outlined by
+// https://github.com/cloud-and-ai-microsoft/resource-provider-contract/blob/master/v1.0/subscription-lifecycle-api-reference.md
+func (s *SubscriptionStateMuxValidator) MiddlewareValidateSubscriptionState(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	subscriptionId := r.PathValue(strings.ToLower(PathSegmentSubscriptionID))
+	sub, exists := s.cache.GetSubscription(subscriptionId)
+	if subscriptionId == "" {
```
Comment (bennerv) [inline frontend/middleware_validatesubscription.go:38]: Shouldn't we check the cache entry exists first before checking the contents of the string?

File: frontend/middleware_validatesubscription.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+			w, http.StatusBadRequest,
+			arm.CloudErrorInvalidSubscriptionState, "",
+			UnregisteredSubscriptionStateMessage,
+			subscriptionId, api.ProviderNamespace)
+	case subscription.Warned:
+		fallthrough // Warned has the same behaviour as Suspended
+	case subscription.Suspended:
+		if r.Method == http.MethodGet || r.Method == http.MethodDelete {
+			next(w, r)
+		} else if r.Method == http.MethodPut || r.Method == http.MethodPatch || r.Method == http.MethodPost {
```
Comment (bennerv) [inline frontend/middleware_validatesubscription.go:72]: can this just be an else?  Should everything else be captured by our mux allowing only specific methods?

**Review by bennerv — COMMENTED** [review]
can we get jira link in PR?

File: frontend/middleware_validatesubscription.go (hunk 6 above)
Comment (bennerv) [reply frontend/middleware_validatesubscription.go:17]: My line of thinking: 
- We should consolidate these somewhere at some point, but not now.  
- BU should be good with error messages.  If they're the same, this is good.

File: frontend/middleware_validatesubscription.go (hunk 3 above)
Comment (s-amann) [reply frontend/middleware_validatesubscription.go:36]: No. At least from what I see in the debugger, all of the keys for PathValue (r.pat.segments) are lowercased, as well as their matched value.

File: frontend/middleware_validatesubscription.go (hunk 7 above)
Comment (s-amann) [reply frontend/middleware_validatesubscription.go:38]: I think you mean, we should check if the string is empty, before checking if the string is key for our cache. I'll fix this.

File: frontend/middleware_validatesubscription.go (hunk 6 above)
Comment (s-amann) [reply frontend/middleware_validatesubscription.go:17]: They are not the same. I can take a look at classic and align with them for now.

File: frontend/middleware_validatesubscription.go (hunk 8 above)
Comment (s-amann) [reply frontend/middleware_validatesubscription.go:72]: sure. I went back and forth on this, and landed on explicitly looking at what the documentation accounts for.

File: frontend/middleware_validatesubscription.go (hunk 7 above)
Comment (bennerv) [reply frontend/middleware_validatesubscription.go:38]: Ah yeah I misread.  

- Check string empty
- fetch from cache
- check if cache miss
- ...

File: frontend/middleware_validatesubscription.go (hunk 6 above)
Comment (bennerv) [reply frontend/middleware_validatesubscription.go:17]: would a `//todo align with aro classic` be a good way to capture these items temporarily for future consolidation?  Or not worth the effort?

Comment (s-amann) [reply frontend/middleware_validatesubscription.go]: might be best to do this type of consolidation when we have more errors to align, for now I updated the ones that do exist in both to match classic.

File: frontend/middleware_validatesubscription.go (hunk 7 above)
Comment (s-amann) [reply frontend/middleware_validatesubscription.go:35]: this is now done.

**Review by mbarnes — COMMENTED** [review]
Couple questions about API versions and the subscription endpoint:

- Can we remove `/internal/api/subscription/register.go`?
I was going to suggest using a different middleware chain for this endpoint that omits `MiddlewareValidateAPIVersion` (since that's only meant for _our_ API versions) but it looks like you've effectively done that already.

- Should `ArmSubscriptionAction` be verfiying the `api-version` parameter is "2.0"?

**Comment by s-amann** [comment]
> Couple questions about API versions and the subscription endpoint:
> 
> * Can we remove `/internal/api/subscription/register.go`?
>   I was going to suggest using a different middleware chain for this endpoint that omits `MiddlewareValidateAPIVersion` (since that's only meant for _our_ API versions) but it looks like you've effectively done that already.

I can delete `subscription/register.go` - I only needed this before removing the endpoint from our api version validation.

> * Should `ArmSubscriptionAction` be verfiying the `api-version` parameter is "2.0"?

I'm not sure about this, the docs say "system version 2.0" will be used, but it doesn't say anything about requiring 2.0

**Review by mbarnes — DISMISSED** [review]
LGTM.  :+1: on moving subscription structs into the `arm` package.

**Review by bennerv — APPROVED** [review]

**Review by mjlshen — APPROVED** [review]

**Pull request merged**